### ISSUE TITLE
forwardIn() with port=0 (dynamic) doesn't work

### ISSUE THREAD

**Issue by dbushong** [mscdex/ssh2#470]
What works:

* `ssh -R '*:0:localhost:8000' my-host` works fine (dynamic port on remote host properly proxies to port 8000 locally)
* The given `forwardIn()` example code, when the port is explicitly specified as shown

What doesn't work:

I modified the `forwardIn()` example given minimally to use the port=0 feature for selecting a dynamic remote port:

```javascript
var Client = require('ssh2').Client;

var conn = new Client();
conn.on('ready', function() {
  console.log('Client :: ready');
  conn.forwardIn('127.0.0.1', 0, function(err, port) {
    if (err) throw err;
    console.log('Listening for connections on server on port ' + port + '!');
  });
}).on('tcp connection', function(info, accept, reject) {
  console.log('TCP :: INCOMING CONNECTION:');
  console.dir(info);
  accept().on('close', function() {
    console.log('TCP :: CLOSED');
  }).on('data', function(data) {
    console.log('TCP :: DATA: ' + data);
  }).end([
    'HTTP/1.1 404 Not Found',
    'Date: Thu, 15 Nov 2012 02:07:58 GMT',
    'Server: ForwardedConnection',
    'Content-Length: 0',
    'Connection: close',
    '',
    ''
  ].join('\r\n'));
}).connect({
  host: 'my-host',
  port: 22,
  username: 'dbushong',
  agent: process.env.SSH_AUTH_SOCK,
});
```

Only changes are:

* replace port 8000 with 0
* take `port` as 2nd arg to callback and report it
* set my username, host, and agent (all 3 worked fine with explicit port)

Result is:

```
$ node bar.js
Client :: ready
Listening for connections on server on port 42469!
```

then in another shell:

```
$ curl -kisS http://my-host:42469/
curl: (52) Empty reply from server
```

(and no further output in the first shell)

I expected the same result as when I did the explicit port:

```
$ curl -kisS http://my-host:8000/
HTTP/1.1 404 Not Found
Date: Thu, 15 Nov 2012 02:07:58 GMT
Server: ForwardedConnection
Content-Length: 0
Connection: close

```

**Comment by mscdex** [comment]
Your `ssh` command line isn't equivalent to your code since the former is asking the server to listen on *all* interfaces, whereas your code is asking the server to only listen on localhost.

With that in mind, does `ssh -R '127.0.0.1:0:localhost:8000' my-host` work?

What about using `'0.0.0.0'` instead of `'127.0.0.1'` in your code for the remote binding address?

**Comment by dbushong** [comment]
So I omitted that part lest it complicated it, but yes, I tried with 0.0.0.0 to the same effect.  The curious part is that when I did the explicit port, but left it at 127.0.0.1, it still allowed me to connect from another host (i.e. despite the 127.0.0.1 it was *not* only listening on the local interface)

**Comment by mscdex** [comment]
What about the other `ssh -R` command?

**Comment by dbushong** [comment]
Additional data point: if I start up `node bar.js`, then ssh into `my-host`, then `curl -kisS http://localhost:$port/` it get the exact same `Empty reply from server`

Re: ssh cli: interestingly, the host seems to serve things remotely even when I put `ssh -R '127.0.0.1:0:localhost:8000' my-host` (yes, this also works, btw)

**Comment by mscdex** [comment]
Can you set `debug: console.log` in the ssh2 connection config and post the output?

Also, could you double check the assigned port with `netstat -nlp` to see if the port is actually bound and to which address?

**Comment by dbushong** [comment]
Here's the debug:

```
DEBUG: Local ident: 'SSH-2.0-ssh2js0.1.8'
DEBUG: Client: Trying my-host on port 22 ...
DEBUG: Client: Connected
DEBUG: Parser: IN_INIT
DEBUG: Parser: IN_GREETING
DEBUG: Parser: IN_HEADER
DEBUG: Remote ident: 'SSH-2.0-OpenSSH_5.3'
DEBUG: Outgoing: Writing KEXINIT
DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: pktLen:836,padLen:10,remainLen:832
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXINIT
DEBUG: Comparing KEXINITs ...
DEBUG: (local) KEX algorithms: ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha1
DEBUG: (remote) KEX algorithms: diffie-hellman-group-exchange-sha256,diffie-hellman-group-exchange-sha1,diffie-hellman-group14-sha1,diffie-hellman-group1-sha1
DEBUG: KEX algorithm: diffie-hellman-group-exchange-sha256
DEBUG: (local) Host key formats: ssh-rsa
DEBUG: (remote) Host key formats: ssh-rsa,ssh-dss
DEBUG: Host key format: ssh-rsa
DEBUG: (local) Client->Server ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com
DEBUG: (remote) Client->Server ciphers: aes128-ctr,aes192-ctr,aes256-ctr,arcfour256,arcfour128,aes128-cbc,3des-cbc,blowfish-cbc,cast128-cbc,aes192-cbc,aes256-cbc,arcfour,rijndael-cbc@lysator.liu.se
DEBUG: Client->Server Cipher: aes128-ctr
DEBUG: (local) Server->Client ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com
DEBUG: (remote) Server->Client ciphers: aes128-ctr,aes192-ctr,aes256-ctr,arcfour256,arcfour128,aes128-cbc,3des-cbc,blowfish-cbc,cast128-cbc,aes192-cbc,aes256-cbc,arcfour,rijndael-cbc@lysator.liu.se
DEBUG: Server->Client Cipher: aes128-ctr
DEBUG: (local) Client->Server HMAC algorithms: hmac-sha2-256,hmac-sha2-512,hmac-sha1
DEBUG: (remote) Client->Server HMAC algorithms: hmac-md5,hmac-sha1,umac-64@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-ripemd160,hmac-ripemd160@openssh.com,hmac-sha1-96,hmac-md5-96
DEBUG: Client->Server HMAC algorithm: hmac-sha2-256
DEBUG: (local) Server->Client HMAC algorithms: hmac-sha2-256,hmac-sha2-512,hmac-sha1
DEBUG: (remote) Server->Client HMAC algorithms: hmac-md5,hmac-sha1,umac-64@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-ripemd160,hmac-ripemd160@openssh.com,hmac-sha1-96,hmac-md5-96
DEBUG: Server->Client HMAC algorithm: hmac-sha2-256
DEBUG: (local) Client->Server compression algorithms: none,zlib@openssh.com,zlib
DEBUG: (remote) Client->Server compression algorithms: none,zlib@openssh.com
DEBUG: Client->Server compression algorithm: none
DEBUG: (local) Server->Client compression algorithms: none,zlib@openssh.com,zlib
DEBUG: (remote) Server->Client compression algorithms: none,zlib@openssh.com
DEBUG: Server->Client compression algorithm: none
DEBUG: Outgoing: Writing KEXDH_GEX_REQUEST
DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: pktLen:404,padLen:8,remainLen:400
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXDH_GEX_GROUP
DEBUG: Outgoing: Writing KEXDH_GEX_INIT
DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: pktLen:956,padLen:9,remainLen:952
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXDH_GEX_REPLY
DEBUG: Checking host key format
DEBUG: Checking signature format
DEBUG: Verifying host fingerprint
DEBUG: Host accepted by default (no verification)
DEBUG: Verifying signature
DEBUG: Outgoing: Writing NEWKEYS
DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: pktLen:12,padLen:10,remainLen:8
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: IN_PACKETDATAAFTER, packet: NEWKEYS
DEBUG: Outgoing: Writing SERVICE_REQUEST (ssh-userauth)
DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: Decrypting
DEBUG: Parser: pktLen:28,padLen:10,remainLen:16
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: Decrypting
DEBUG: Parser: HMAC size:32
DEBUG: Parser: IN_PACKETDATAVERIFY
DEBUG: Parser: Verifying MAC
DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
DEBUG: Parser: IN_PACKETDATAAFTER, packet: SERVICE_ACCEPT
DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
DEBUG: Outgoing: Writing USERAUTH_REQUEST (publickey -- check)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: Decrypting
DEBUG: Parser: pktLen:300,padLen:4,remainLen:288
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: Decrypting
DEBUG: Parser: HMAC size:32
DEBUG: Parser: IN_PACKETDATAVERIFY
DEBUG: Parser: Verifying MAC
DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
DEBUG: Parser: IN_PACKETDATAAFTER, packet: USERAUTH_PK_OK
DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
DEBUG: Outgoing: Writing USERAUTH_REQUEST (publickey)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: Decrypting
DEBUG: Parser: pktLen:12,padLen:10,remainLen:0
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: HMAC size:32
DEBUG: Parser: IN_PACKETDATAVERIFY
DEBUG: Parser: Verifying MAC
DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
DEBUG: Parser: IN_PACKETDATAAFTER, packet: USERAUTH_SUCCESS
Client :: ready
DEBUG: Outgoing: Writing GLOBAL_REQUEST (tcpip-forward)
DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: Decrypting
DEBUG: Parser: pktLen:12,padLen:6,remainLen:0
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: HMAC size:32
DEBUG: Parser: IN_PACKETDATAVERIFY
DEBUG: Parser: Verifying MAC
DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
DEBUG: Parser: IN_PACKETDATAAFTER, packet: REQUEST_SUCCESS
Listening for connections on server on port 23720!
DEBUG: Parser: IN_PACKETBEFORE (expecting 16)

(here is where I ran `curl -kisS http://my-host:23720/`)

DEBUG: Parser: IN_PACKET
DEBUG: Parser: Decrypting
DEBUG: Parser: pktLen:76,padLen:6,remainLen:64
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: Decrypting
DEBUG: Parser: HMAC size:32
DEBUG: Parser: IN_PACKETDATAVERIFY
DEBUG: Parser: Verifying MAC
DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_OPEN (2, forwarded-tcpip)
DEBUG: Client: Automatic rejection of incoming channel open: unexpected channel open for: forwarded-tcpip
DEBUG: Outgoing: Writing CHANNEL_OPEN_FAILURE (2)
DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
```

and here's the relevant lines from `sudo netstat -nlp` on my-host:

```
tcp        0      0 0.0.0.0:23720               0.0.0.0:*                   LISTEN      4725/sshd
tcp        0      0 :::23720                    :::*                        LISTEN      4725/sshd
```

**Comment by dbushong** [comment]
ah hah, found it:
https://github.com/mscdex/ssh2/blob/master/lib/client.js#L1443

that's comparing destPort 0 vs destPort 23720

I'll try to poke around and make a PR if you don't get to it first.

**Comment by mscdex** [comment]
What port is the server sending for the incoming connection? It should be the bound port and not 0.

**Comment by dbushong** [comment]
It is sending the correct port, but AFAICT the only place that ever gets used is here:
https://github.com/mscdex/ssh2/blob/master/lib/client.js#L918-L923

and the `info.data` available to `onCHANNEL_OPEN()` doesn't have that port updated.

**Comment by mscdex** [comment]
I believe this may be a bug with OpenSSH 5.x. Later OpenSSH versions have a remote port bug flag for those versions.

**Comment by dbushong** [comment]
Ah hah.  ;;  This doesn't seem *too* hard to workaround (I don't have the option of upgrading SSH on the server)  Maybe I'll make a PR that documents the issue in the README section about port=0 or offers a workaround.  Feel free to close this issue.

**Comment by mscdex** [comment]
I think I will just implement the same workaround that OpenSSH does, namely ignoring the port number returned at bind time for OpenSSH 5.x when listening on port 0 (thus just listening for `<address>:0`).

**Comment by dbushong** [comment]
:+1: thx

**Comment by mscdex** [comment]
Fix should now be published in v0.5.2.

**Comment by dbushong** [comment]
Works!  Thanks a bunch!
